### ISSUE TITLE
I have a problem with "stash" in the windows version

### ISSUE THREAD

**Issue by lyingfromyou** [git-for-windows/git#1579]
like this :
When I execute "stash" on the "dev" branch, the changed file still exists in the workspace
![1](https://user-images.githubusercontent.com/24619864/37823582-de9b2a1a-2ec4-11e8-96e6-e1338d0e9eda.png)
![2](https://user-images.githubusercontent.com/24619864/37823583-df4f30dc-2ec4-11e8-898a-f9f552d7bfa7.png)
![3](https://user-images.githubusercontent.com/24619864/37823586-e0a4c5be-2ec4-11e8-8bd9-3460bbc5f7d0.png)
![4](https://user-images.githubusercontent.com/24619864/37823592-e22ce664-2ec4-11e8-8a21-abb5b99edd45.png)

**Comment by lyingfromyou** [comment]
git version 2.16.0.windows.2

**Comment by dscho** [comment]
Please do paste text, not screenshots. Now I have to type out that long path myself.

What does

```sh
git ls-files --eol src/main/java/com/lying/docking/request1/CoreRequestUtil.java
```

say? (I suspect line ending issues to be at the core of the problem, i.e. `CoreRequestUtil.java` might have been committed with CR/LF line endings instead of LF-only as Git would expect.)

**Comment by lyingfromyou** [comment]
I have confirmed that the end of the line is CR/LF.

**Comment by lyingfromyou** [comment]
I have confirmed that the end of the line is CR/LF.
my operation is as follows:

    $: git checkout -b dev
    $: # modify a file that has been pushed
    $: vi xxx   
    $: # after modifying
    $: git stash

like I said above, the changed file still exists in the workspace.

**Comment by dscho** [comment]
And what does that `ls-files` command say?

**Comment by lyingfromyou** [comment]
After "git stash":
"git ls-files -m " and "git ls-files -s"
The file can be seen in both commands

**Comment by dscho** [comment]
@lyingfromyou I meant the output of `git ls-files --eol src/main/java/com/lying/docking/request1/CoreRequestUtil.java`. You still have not imparted that information.

**Comment by dscho** [comment]
Ping? @lyingfromyou are you still interested in figuring this one out?

**Comment by cincauhangus** [comment]
I'm having similar issues as @lyingfromyou and using the same version. I've originally posted a question on [SO](https://stackoverflow.com/questions/51188270/git-stash-doesnt-stash-changes-properly) before coming here. 

When I run `git ls-files --eol` on one of the files, this is the output:
```
$ git ls-files --eol app/Classes/ModelExtension.php
i/lf    w/lf    attr/text eol=lf        app/Classes/ModelExtension.php
```

As reference, my `.gitattributes` file defines the following:
```
*          text=auto
*.php      text eol=lf
```

**Comment by tboegi** [comment]
Your line 
>i/lf    w/lf    attr/text eol=lf        app/Classes/ModelExtension.php
Looks correct to me so far.

Could you elaborate which problem you exactly have?
(Saying similar too ... is a good start, but it may be easier
 to start the description from the beginning)

**Comment by cincauhangus** [comment]
Sure @tboegi!

I'm running `git version 2.16.0.windows.2`. Whenever I run `git stash` on any of my repos, git will respond with usual confirmation:

```
$ git stash
Saved working directory and index state WIP on master: 0000000 changes
```

After that, I run `git status -uno` and I'll still see the staged and unstaged changes as if I've not run the `git stash` command:

```
$ git status -uno
On branch master

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   app/Classes/ModelExtension.php

no changes added to commit (use "git add" and/or "git commit -a")
```

`git stash show 0` will return the list of stashed files, as the same output from the `git status -uno` command:

```
$ git stash list
stash@{0}: WIP on master: 0000000 changes

$ git stash show 0
 app/Classes/ModelExtension.php | 1 +
 1 file changed, 1 insertion(+)
```

From here, running `git stash` again will remove the staged and unstaged changes from the repo. `git status -uno` also confirms with the "nothing to commit" message:


```
$ git stash
Saved working directory and index state WIP on master: 0000000 changes

$ git status -uno
On branch master

nothing to commit (use -u to show untracked files)

$ git stash list
stash@{0}: WIP on master: 0000000 changes
stash@{1}: WIP on master: 0000000 changes
```

Running `git stash show 0` now shows no output (an empty stash?), and `git stash show 1` returns the list of stashed items:

```
$ git stash show 0 #no output here

$ git stash show 1
 app/Classes/ModelExtension.php | 1 +
 1 file changed, 1 insertion(+)
```

When I want to unstash the changes, I'll need to run `git stash pop` twice to get to correct stash reference. First pop will return "Already up to date!" and the second pop will unstash my changes:

```
$ git stash pop
Already up to date!
On branch master

nothing added to commit 
Dropped refs/stash@{0} (299485c1940df276b19eae55b90595b5256912fa)

$ git stash pop
On branch master

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   app/Classes/ModelExtension.php

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (43135a4e20f9d7b5604b98114f634bff642e4ff2)
```

**Comment by tboegi** [comment]
The content of your repo is impossible to debug from here.
(Or is it a public repo?)

Can you run
git ls-files --eol -s | grep -i app/Classes/ModelExtension.php

**Comment by cincauhangus** [comment]
Private repo, unfortunately.

> Can you run
git ls-files --eol -s | grep -i app/Classes/ModelExtension.php

```
100644 df7893ab6fe44b19986f1e22ddc99689cf3e8491 0       i/lf    w/lf    attr/text eol=lf        app/Classes/ModelExtension.php
```

**Comment by dscho** [comment]
> Private repo, unfortunately

@cincauhangus So basically you are asking others to come up with a fresh repository/worktree that exposes the same problems.

How about tipping the burden a little in your direction, and you start a new one, from scratch, trying to replicate the issue on your side? Once you can replicate the bug(s), you can publish *that* repository. That way, if the problem is purely in your setup, you won't have wasted the hours of somebody like e.g. me who would simply fail to reproduce the problems due to those differences in the setup.

In other words, do try to make the time (as you wish helpful souls to make the time to help you!) to come up with an [MCVE](https://stackoverflow.com/help/mcve).

**Comment by cincauhangus** [comment]
Many apologies. I've created a new public repo with a single file and I'm able to reproduce this behaviour on my machine. I've also included the gitattributes file. 

https://github.com/cincauhangus/git-4-windows-issue-1579

With the contents of the repo above, I made a change and stashed it using the same steps defined [here](https://github.com/git-for-windows/git/issues/1579#issuecomment-402696979)

Something to note, my editor (VS Code) created the file in `crlf` mode, and gitattribute defines .php is to be saved in `lf`.
